ambry-frontend/src/main/java/com/github/ambry/frontend/ListDatasetVersionHandler.java

```diff
@@ -0,0 +1,162 @@
+    new ListDatasetVersionHandler.CallbackChain(restRequest, restResponseChannel, callback).start();
+  }
+
+  /**
+   * Represents the chain of actions to take. Keeps request context that is relevant to all callback stages.
+   */
+  private class CallbackChain {
+    private final RestRequest restRequest;
+    private final String uri;
+    private final RestResponseChannel restResponseChannel;
+    private final Callback<ReadableStreamChannel> finalCallback;
+
+    /**
+     * @param restRequest the {@link RestRequest}.
+     * @param restResponseChannel the {@link RestResponseChannel}.
+     * @param finalCallback the {@link Callback} to call on completion.
+     */
+    private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<ReadableStreamChannel> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.finalCallback = finalCallback;
+      this.uri = restRequest.getUri();
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#preProcessRequest}.
+     */
+    private void start() {
+      // Start the callback chain by performing request security processing.
+      securityService.processRequest(restRequest, securityProcessRequestCallback());
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(frontendMetrics.getDatasetsSecurityProcessRequestMetrics,
+          securityCheckResult -> securityService.postProcessRequest(restRequest, securityPostProcessRequestCallback()),
+          uri, LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityPostProcessRequestCallback() {
+      return buildCallback(frontendMetrics.getDatasetsSecurityPostProcessRequestMetrics, securityCheckResult -> {
+        LOGGER.debug("Received request for listing all datasets with arguments: {}", restRequest.getArgs());
+        accountAndContainerInjector.injectDatasetForNamedBlob(restRequest);
+        Page<String> datasetList = listAllValidDatasetVersions();
+        ReadableStreamChannel channel = FrontendUtils.serializeJsonToChannel(
+            datasetList.toJson(datasetVersion -> new JSONObject().put(DATASET_VERSION_KEY, datasetVersion)));
```

Comment:
do we need to return a list of object that only contains one version key back to client? Can we just return a list of versions?

Reply:
Updated to use new serialize and deserialize logic in Page.